Annotations

pese-git opened the pull request.

pese-git committed:
feat: Add async dependency resolution and enhance example

- Implemented async provider methods `toProvideAsync` and `toProvideAsyncWithParams` in `Binding` class, allowing asynchronous initialization with dynamic parameters.
- Added typedefs `AsyncProvider<T>` and `AsyncProviderWithParams<T>` for better type clarity with async operations.
- Introduced async resolution methods `resolveAsync` and `tryResolveAsync` in `Scope` for resolving asynchronous dependencies.
- Updated example in `main.dart` to demonstrate async dependency resolution capabilities.
  - Modified `FeatureModule` to utilize async providers for `DataRepository` and `DataBloc`.
  - Replaced synchronous resolution with `resolveAsync` where applicable.
  - Handled potential errors in dependency resolution with try-catch.
- Removed unnecessary whitespace for cleaner code formatting.

pese-git committed:
feat: implement annotations

pese-git committed:
start implement generator code

pese-git committed:
feat: implement generator di module

pese-git committed:
fix: fix generator for  singletone annotation

pese-git committed:
feat: implement generator for named annotation

pese-git committed:
hotfix

pese-git committed:
feat: implement named dependency

pese-git committed:
fix: fix module generator

pese-git committed:
modify sample

pese-git committed:
write comments to code

pese-git committed:
doc: added comments to code

pese-git committed:
fix: fix warnings

pese-git committed:
doc: update README

pese-git committed:
doc: add README

pese-git committed:
doc: update readme

pese-git committed:
doc: update readme

pese-git committed:
Merge pull request #7 from pese-git/develop

pese-git committed:
doc: fix comment

pese-git committed:
chore(release): publish packages

 - cherrypick@2.1.0
 - cherrypick_flutter@1.1.1

pese-git committed:
fix readme and freez deps

pese-git committed:
add provide typedef

pese-git committed:
feat: implement toInstanceAync binding

pese-git committed:
implement scope tests

pese-git committed:
update tests

pese-git committed:
Merge pull request #8 from pese-git/refactor

Refactor code and add toInstanceAsync method

pese-git committed:
update readme

pese-git committed:
refactor module generator

pese-git committed:
feat: implement instance/provide annotations

pese-git committed:
feat: generate instance async code

pese-git committed:
feat: implement async mode for instance/provide annotations

pese-git committed:
feat: implement generator for dynamic params

pese-git committed:
fix: fix warning conflict with names

pese-git committed:
refactor code

pese-git committed:
refactor code

pese-git committed:
added documentations

pese-git committed:
update readme

pese-git committed:
doc: update documentations

pese-git committed:
doc: update documentations

pese-git committed:
doc: update readme

pese-git committed:
fix pubspecs

pese-git committed:
chore(release): publish packages

 - cherrypick@2.2.0-dev.0
 - cherrypick_annotations@1.1.0-dev.0
 - cherrypick_flutter@1.1.2-dev.0
 - cherrypick_generator@1.1.0-dev.0

pese-git committed:
fix pubspecs

KlimYarosh commented on cherrypick_generator/lib/src/bind_spec.dart:
```diff
@@ -0,0 +1,245 @@
+    // Safe variable name for parameters.
+    const paramVar = 'args';
+    final fnArgs = parameters
+        .map((p) => p.isParams ? paramVar : p.generateArg(paramVar))
```
```suggestion
        .map((p) => p.generateArg(paramVar))
```

KlimYarosh replied:
проверка реализована в методе

pese-git replied:
done

KlimYarosh commented on cherrypick_generator/lib/src/bind_spec.dart:
```diff
@@ -0,0 +1,245 @@
+  final List<BindParameterSpec> parameters;
+
+  /// Binding type: 'instance' or 'provide' (@instance or @provide)
+  final String bindingType; // 'instance' | 'provide'
```
Чтобы не допустить опечатки в написании типов либо добавить статики названий типов, либо сделать расширенный enum с полем value к примеру.

pese-git replied:
Добавил enum

done

KlimYarosh commented on cherrypick_generator/lib/src/bind_spec.dart:
```diff
@@ -0,0 +1,245 @@
+    switch (bindingType) {
+      case 'instance':
+        return isAsyncInstance
+            ? '.toInstanceAsync(($fnArgs) => $methodName($fnArgs))'
+            : '.toInstance(($fnArgs) => $methodName($fnArgs))';
```
сигнатура метода иная
`'.toInstanceAsync($methodName($argsStr))'`
`'. toInstance($methodName($argsStr))'`

то есть по сути, если мы указываем, что у нас instance c параметрами, то это
provide с singleton = true.

Вероятно должно быть что-то типа:
```
  String target;
  if (isAsyncInstance) {
          target = multiLine
              ? '.toProvideAsyncWithParams(\n${' ' * (indent + 2)}($paramVar) => $methodName($fnArgs))'
              : '.toProvideAsyncWithParams(($paramVar) => $methodName($fnArgs))';
        } else {
          target =  multiLine
              ? '.toProvideWithParams(\n${' ' * (indent + 2)}($paramVar) => $methodName($fnArgs))'
              : '.toProvideWithParams(($paramVar) => $methodName($fnArgs))';
        }
  return '$target.singleton()';
```

KlimYarosh replied:
P.S. Возможно нужно будет добавить к provide флаг lazy. На будущее.

pese-git replied:
Не совсем. Может быть ситуация, когда один instance зависит от  одного или нескольких других.

```dart
bind<int>().toInstance(1).withName('count');
bind<str>().toInstance('Print: ${currentScope.resolve<int>(named: 'count')}').withName('echo');
```

pese-git replied:
done

pese-git committed:
fix: optimize code

pese-git committed:
refactor code

pese-git committed:
update deps

pese-git committed:
fix: fix warnings

pese-git committed:
chore(release): publish packages

 - cherrypick@2.2.0-dev.1
 - cherrypick_generator@1.1.0-dev.1
 - cherrypick_flutter@1.1.2-dev.1

pese-git committed:
update pubspec

pese-git committed:
update pubspec

pese-git committed:
fix: update instance generator code

pese-git committed:
update deps

pese-git committed:
doc: update documentations

pese-git committed:
chore(release): publish packages

 - cherrypick_generator@1.1.0-dev.2

pese-git committed:
starting implement inject generator

pese-git committed:
feat: implement InjectGenerator

pese-git committed:
implement inject generator

pese-git committed:
refactor inject generator

pese-git committed:
doc: update documentations

pese-git committed:
chore(release): publish packages

 - cherrypick_annotations@1.1.0-dev.1
 - cherrypick_generator@1.1.0-dev.3

pese-git committed:
update documentations

pese-git committed:
update documentaions

pese-git committed:
update documentaions

pese-git committed:
chore: update deps and up to flutter sdk 3.29.3 and dart >=3.7.0

pese-git committed:
fix: fixed warnings

pese-git committed:
chore(release): publish packages

 - cherrypick_generator@1.1.0-dev.4

pese-git committed:
feat: implement tryResolve via generate code

pese-git committed:
chore(release): publish packages

 - cherrypick_generator@1.1.0-dev.5

pese-git committed:
feat(generator): complete code generation testing framework with 100% test coverage

BREAKING CHANGE: Updated file extensions and dependencies for better compatibility

## 🎯 Major Features Added:
- ✅ Complete test suite for ModuleGenerator (66 integration tests)
- ✅ Complete test suite for InjectGenerator (66 integration tests)
- ✅ Comprehensive unit tests for BindSpec, MetadataUtils
- ✅ 195 total tests across all packages (100% passing)

## 🔧 Technical Improvements:
- feat(generator): add comprehensive integration tests for code generation
- feat(generator): implement BindSpec unit tests with full coverage
- feat(generator): add MetadataUtils unit tests for annotation processing
- fix(generator): update file extensions to avoid conflicts (.module.cherrypick.g.dart)
- fix(generator): correct part directive generation in templates
- fix(generator): resolve dart_style 3.x formatting compatibility

## 📦 Dependencies & Configuration:
- build(deps): upgrade analyzer to ^7.0.0 for Dart 3.5+ compatibility
- build(deps): upgrade dart_style to ^3.0.0 for modern formatting
- build(deps): upgrade source_gen to ^2.0.0 for latest features
- config(build): update build.yaml with new file extensions
- config(melos): optimize test commands for better performance

## 🐛 Bug Fixes:
- fix(examples): correct local package paths in client_app and postly
- fix(analysis): exclude generated files from static analysis
- fix(generator): remove unused imports and variables
- fix(tests): add missing part directives in test input files
- fix(tests): update expected outputs to match dart_style 3.x format

## 🚀 Performance & Quality:
- perf(tests): optimize test execution time (132 tests in ~1 second)
- quality: achieve 100% test coverage for code generation
- quality: eliminate all analyzer warnings and errors
- quality: ensure production-ready stability

## 📋 Test Coverage Summary:
- cherrypick: 61 tests ✅
- cherrypick_annotations: 1 test ✅
- cherrypick_generator: 132 tests ✅
- cherrypick_flutter: 1 test ✅
- Total: 195 tests (100% passing)

## 🔄 Compatibility:
- ✅ Dart SDK 3.5.2+
- ✅ Flutter 3.24+
- ✅ melos + fvm workflow
- ✅ build_runner integration
- ✅ Modern analyzer and formatter

This commit establishes CherryPick as a production-ready dependency injection
framework with enterprise-grade testing and code generation capabilities.

pese-git committed:
fix: format test code
